File: backend/internal/service/support.go
package service

import (
	"context"
	"time"

	"github.com/Wei-Shaw/sub2api/internal/domain"
	infraerrors "github.com/Wei-Shaw/sub2api/internal/pkg/errors"
	"github.com/Wei-Shaw/sub2api/internal/pkg/pagination"
)

const (
	SupportTicketStatusOpen         = domain.SupportTicketStatusOpen
	SupportTicketStatusPendingAdmin = domain.SupportTicketStatusPendingAdmin
	SupportTicketStatusPendingUser  = domain.SupportTicketStatusPendingUser
	SupportTicketStatusClosed       = domain.SupportTicketStatusClosed
)

const (
	SupportTicketCategoryGeneral      = domain.SupportTicketCategoryGeneral
	SupportTicketCategoryRecharge     = domain.SupportTicketCategoryRecharge
	SupportTicketCategorySubscription = domain.SupportTicketCategorySubscription
	SupportTicketCategoryAPIIssue     = domain.SupportTicketCategoryAPIIssue
	SupportTicketCategoryAccount      = domain.SupportTicketCategoryAccount
)

const (
	SupportTicketPriorityLow    = domain.SupportTicketPriorityLow
	SupportTicketPriorityNormal = domain.SupportTicketPriorityNormal
	SupportTicketPriorityHigh   = domain.SupportTicketPriorityHigh
	SupportTicketPriorityUrgent = domain.SupportTicketPriorityUrgent
)

const (
	SupportMessageSenderUser   = domain.SupportMessageSenderUser
	SupportMessageSenderAdmin  = domain.SupportMessageSenderAdmin
	SupportMessageSenderSystem = domain.SupportMessageSenderSystem
)

var (
	ErrSupportTicketNotFound       = domain.ErrSupportTicketNotFound
	ErrSupportInputRequired        = infraerrors.BadRequest("SUPPORT_INPUT_REQUIRED", "support input is required")
	ErrSupportTitleInvalid         = infraerrors.BadRequest("SUPPORT_TITLE_INVALID", "support ticket title is invalid")
	ErrSupportContentInvalid       = infraerrors.BadRequest("SUPPORT_CONTENT_INVALID", "support ticket message content is invalid")
	ErrSupportCategoryInvalid      = infraerrors.BadRequest("SUPPORT_CATEGORY_INVALID", "support ticket category is invalid")
	ErrSupportStatusInvalid        = infraerrors.BadRequest("SUPPORT_STATUS_INVALID", "support ticket status is invalid")
	ErrSupportPriorityInvalid      = infraerrors.BadRequest("SUPPORT_PRIORITY_INVALID", "support ticket priority is invalid")
	ErrSupportSenderRoleInvalid    = infraerrors.BadRequest("SUPPORT_SENDER_ROLE_INVALID", "support ticket sender role is invalid")
	ErrSupportTicketClosed         = infraerrors.Conflict("SUPPORT_TICKET_CLOSED", "support ticket is closed")
	ErrSupportTicketAlreadyOpen    = infraerrors.Conflict("SUPPORT_TICKET_ALREADY_OPEN", "support ticket is not closed")
	ErrSupportRateLimited          = infraerrors.TooManyRequests("SUPPORT_RATE_LIMITED", "too many support requests")
	ErrSupportAssignedAdminInvalid = infraerrors.BadRequest("SUPPORT_ASSIGNED_ADMIN_INVALID", "assigned admin id is invalid")
)

type SupportTicket = domain.SupportTicket
type SupportTicketMessage = domain.SupportTicketMessage
type SupportTicketUser = domain.SupportTicketUser

type SupportTicketListFilters struct {
	UserID     int64
	Status     string
	Category   string
	Priority   string
	Search     string
	UnreadOnly bool
}

type SupportMessageListFilters struct {
	BeforeID int64
	Limit    int
}

type SupportTicketRepository interface {
	CreateWithMessage(ctx context.Context, ticket *SupportTicket, message *SupportTicketMessage) error
	GetByID(ctx context.Context, id int64) (*SupportTicket, error)
	GetByIDForUser(ctx context.Context, userID, id int64) (*SupportTicket, error)
	List(ctx context.Context, params pagination.PaginationParams, filters SupportTicketListFilters) ([]SupportTicket, *pagination.PaginationResult, error)
	Update(ctx context.Context, ticket *SupportTicket) error
	UpdateReadAt(ctx context.Context, ticketID int64, role string, readAt time.Time) error
	CountCreatedSince(ctx context.Context, userID int64, since time.Time) (int, error)
	CountMessagesSince(ctx context.Context, userID int64, since time.Time) (int, error)
}

type SupportTicketMessageRepository interface {
	Create(ctx context.Context, ticket *SupportTicket, message *SupportTicketMessage) error
	ListByTicketID(ctx context.Context, ticketID int64, filters SupportMessageListFilters) ([]SupportTicketMessage, error)
}

File: frontend/src/api/admin/support.ts
import { apiClient } from '../client'
import type {
  BasePaginationResponse,
  SupportMessageListFilters,
  SupportTicket,
  SupportTicketMessage,
  UpdateSupportTicketRequest
} from '@/types'

export async function list(
  page = 1,
  pageSize = 20,
  filters?: {
    status?: string
    category?: string
    priority?: string
    search?: string
    unread_only?: boolean
  }
): Promise<BasePaginationResponse<SupportTicket>> {
  const { data } = await apiClient.get<BasePaginationResponse<SupportTicket>>('/admin/support/tickets', {
    params: { page, page_size: pageSize, ...filters }
  })
  return data
}

export async function getById(id: number): Promise<SupportTicket> {
  const { data } = await apiClient.get<SupportTicket>(`/admin/support/tickets/${id}`)
  return data
}

export async function listMessages(
  id: number,
  filters?: SupportMessageListFilters
): Promise<SupportTicketMessage[]> {
  const { data } = await apiClient.get<SupportTicketMessage[]>(`/admin/support/tickets/${id}/messages`, {
    params: filters
  })
  return data
}

export async function sendMessage(id: number, content: string): Promise<SupportTicketMessage> {
  const { data } = await apiClient.post<SupportTicketMessage>(`/admin/support/tickets/${id}/messages`, { content })
  return data
}

export async function markRead(id: number): Promise<{ message: string }> {
  const { data } = await apiClient.post<{ message: string }>(`/admin/support/tickets/${id}/read`)
  return data
}

export async function update(id: number, request: UpdateSupportTicketRequest): Promise<SupportTicket> {
  const { data } = await apiClient.put<SupportTicket>(`/admin/support/tickets/${id}`, request)
  return data
}

export async function close(id: number): Promise<SupportTicket> {
  const { data } = await apiClient.post<SupportTicket>(`/admin/support/tickets/${id}/close`)
  return data
}

export async function reopen(
  id: number,
  content = ''
): Promise<{ ticket: SupportTicket; message?: SupportTicketMessage | null }> {
  const { data } = await apiClient.post<{ ticket: SupportTicket; message?: SupportTicketMessage | null }>(
    `/admin/support/tickets/${id}/reopen`,
    { content }
  )
  return data
}

const supportAPI = {
  list,
  getById,
  listMessages,
  sendMessage,
  markRead,
  update,
  close,
  reopen
}

export default supportAPI
